usr/share/rear/layout/save/default/950_verify_disklayout_file.sh

```diff
@@ -47,21 +47,39 @@ while read keyword disk_dev disk_size parted_mklabel junk ; do
     # as above layout/prepare/GNU/Linux/100_include_partition_code.sh is the most important one
     # so that it is used here as reference to decide whether or not the entries are correct:
     partitions=()
-    while read keyword dummy part_size part_start part_name part_flags part_dev junk ; do
-        test -b "$part_dev" || broken_part_errors=( "${broken_part_errors[@]}" "$part_dev is not a block device" )
-        is_positive_integer $part_size || broken_part_errors=( "${broken_part_errors[@]}" "$part_dev size $part_size is not a positive integer" )
-        is_nonnegative_integer $part_start || broken_part_errors=( "${broken_part_errors[@]}" "$part_dev start $part_start is not a nonnegative integer" )
-        partitions=( "${partitions[@]}" "$part_dev" )
-        # Using the parted_mklabel fallback behaviour in create_partitions() in prepare/GNU/Linux/100_include_partition_code.sh
-        # only when there is no parted_mklabel value, but when there is a parted_mklabel value use it as is:
-        if ! test "$parted_mklabel" ; then
-            case $part_name in
-                (primary|extended|logical)
-                    parted_mklabel="msdos"
-                    ;;
-            esac
-        fi
-    done < <( grep "^part $disk_dev " "$DISKLAYOUT_FILE" )
+   
+    # cf. https://github.com/rear/rear/issues/2137
+    # s390 - partitions do not list the boot flag
+    # flag is removed from while read ... - this is true for sles also
+    # example
+    # /dev/dasda:24.2GB:dasd:512:4096:dasd:IBM S390 DASD drive:;
+    # 1:98.3kB:524MB:524MB:xfs::;
+    # 2:524MB:24.2GB:23.6GB:::lvm;
+    if [ "$ARCH" == "Linux-s390" ] ; then
+	 while read keyword dummy part_size part_start part_flags part_dev junk ; do
+            test -b "$part_dev" || broken_part_errors=( "${broken_part_errors[@]}" "$part_dev is not a block device" )
```

Comment:
I think it will cause endless trouble at various different places
when you try to make things work `if [ "$ARCH" == "Linux-s390" ]`
without a non-empty `<partition type|name>` value in disklayout.conf

My reasoning is that the `part` entries in disklayout.conf are evaluated
at various different places in various scripts so that the `part` entries
must match the syntax as defined in the `Disk layout file syntax` section in
https://github.com/rear/rear/blob/master/doc/user-guide/06-layout-configuration.adoc
which reads in particular
```
part <disk name> <size(B)> <start(B)> <partition name/type> <flags/"none"> <partition name>
```

I experienced the empty `<partition type|name>` value in disklayout.conf
when I did my first attempts to run "rear mkrescue" on s390.

To get a non-empty `<partition type|name>` value in disklayout.conf
I made the following quick changes on my test system:
```
--- usr/share/rear/layout/save/GNU/Linux/200_partition_layout.sh.original       2019-04-29 17:42:53.039143520 +0200
+++ usr/share/rear/layout/save/GNU/Linux/200_partition_layout.sh        2019-04-29 18:21:13.289143520 +0200
@@ -140,7 +140,7 @@
     ### Find partition name for GPT disks.
     # For the SUSE specific gpt_sync_mbr partitioning scheme
     # see https://github.com/rear/rear/issues/544
-    if [[ "$disk_label" = "gpt" || "$disk_label" == "gpt_sync_mbr" ]] ; then
+    if [[ "$disk_label" = "gpt" || "$disk_label" == "gpt_sync_mbr" || "$disk_label" == "dasd" ]] ; then
         if [[ "$FEATURE_PARTED_MACHINEREADABLE" ]] ; then
             while read partition_nr size start junk ; do
                 # In case of GPT the 'type' field contains actually the GPT partition name.
```
and
```
--- usr/share/rear/layout/prepare/GNU/Linux/100_include_partition_code.sh.original      2019-04-29 17:42:53.039143520 +0200
+++ usr/share/rear/layout/prepare/GNU/Linux/100_include_partition_code.sh       2019-04-29 18:36:16.459143520 +0200
@@ -132,7 +132,7 @@
             ### GPT disks need 33 LBA blocks at the end of the disk
             # For the SUSE specific gpt_sync_mbr partitioning scheme
             # see https://github.com/rear/rear/issues/544
-            if [[ "$label" == "gpt" || "$label" == "gpt_sync_mbr" ]] ; then
+            if [[ "$label" == "gpt" || "$label" == "gpt_sync_mbr" || "$label" == "dasd" ]] ; then
                 device_size=$( mathlib_calculate "$device_size - 33*$block_size" )
                 # Only if resizing all partitions is explicity wanted
                 # resizing of arbitrary partitions may also happen via the code below
```
and
```
--- usr/share/rear/layout/save/default/950_verify_disklayout_file.sh.original   2019-04-29 17:42:53.049143520 +0200
+++ usr/share/rear/layout/save/default/950_verify_disklayout_file.sh    2019-04-29 18:25:16.329143520 +0200
@@ -81,7 +81,7 @@
     # because that indicates partitions of another form than /dev/sdX1 /dev/sdX2 /dev/sdX3 are used:
     if test $highest_used_part_num -gt 0 ; then
         case $parted_mklabel in
-            (gpt)
+            (gpt|dasd)
                 # For the GPT partitioning scheme the partitions must have consecutive numbers 1 2 3 ..
                 non_consecutive_part_found=""
                 unused_part_nums=()
```

I think you should apply at least the first two i.e. the changes
in usr/share/rear/layout/save/GNU/Linux/200_partition_layout.sh
and usr/share/rear/layout/prepare/GNU/Linux/100_include_partition_code.sh

My change in usr/share/rear/layout/save/default/950_verify_disklayout_file.sh
was only a quick hack to make it also working for `dasd`.
I don't know if a `dasd` partitioning scheme matches more a GPT partitioning scheme
or a MBR partitioning scheme.
I let the `dasd` partitioning scheme matche a GPT partitioning scheme
because I think a `dasd` partitioning scheme does not have
overcomplicated things like primary, extended and logical partitions.
Because a `dasd` partitioning scheme is just one partition after the other
(where all partitions are of same kind) it matches more the simplicity of the
GPT partitioning scheme regardless that only 4 `dasd` partitions are allowed.

Reply:
With current changes from this pull request I got this `part` entries in disklayout.conf
```
part /dev/dasda 209682432 98304 none  /dev/dasda1
part /dev/dasda 1026539520 209780736 none  /dev/dasda2
part /dev/dasda 6149013504 1236320256 none  /dev/dasda3
```

Then I reverted the current changes from this pull request in
usr/share/rear/layout/save/default/950_verify_disklayout_file.sh
and appiled my above changes and got `part` entries in disklayout.conf
```
part /dev/dasda 209682432 98304 rear-noname none /dev/dasda1
part /dev/dasda 1026539520 209780736 rear-noname none /dev/dasda2
part /dev/dasda 6149013504 1236320256 rear-noname none /dev/dasda3
```
so that now the `<partition type|name>` value is the `rear-noname` fallback
in usr/share/rear/layout/prepare/GNU/Linux/100_include_partition_code.sh
which likely needs to be enhanced to result `dasd` as that value...

Reply:
When i first went through this, I was on the fence on how to write the part info. I think your approach is better.
I initially went with msdos type, I think it make sense to try the gpd format, as you said, it gives more flexibility

rhel parted is similar:
Model: IBM S390 DASD drive (dasd)
Disk /dev/dasda: 24.2GB
Sector size (logical/physical): 512B/4096B
Partition Table: dasd
Disk Flags: 

```
Number  Start   End     Size    File system  Flags
 1      98.3kB  524MB   524MB   xfs
 2      524MB   24.2GB  23.6GB               lvm
```

I will make the change.